File: repertoire/data-integration/providers/conflict_resolver/crdt_merge.ts
// CRDT-based conflict-free merge resolver
// Always auto-resolves mathematically using convergent replicated data types.
// Applies per-field CRDT merge strategies: LWW-Register for scalars,
// G-Counter for numeric increments, OR-Set for collections.

export interface Conflict {
  entityId: string;
  versionA: Record<string, unknown>;
  versionB: Record<string, unknown>;
  ancestor?: Record<string, unknown>;
  fieldConflicts: string[];
  timestampA?: number;
  timestampB?: number;
}

export interface ResolverConfig {
  options?: Record<string, unknown>;
}

export interface Resolution {
  winner: Record<string, unknown>;
  strategy: string;
  details: Record<string, unknown>;
}

export const PROVIDER_ID = 'crdt_merge';
export const PLUGIN_TYPE = 'conflict_resolver';

type CrdtStrategy = 'lww_register' | 'g_counter' | 'or_set';

function detectFieldCrdtType(valA: unknown, valB: unknown, ancestor: unknown): CrdtStrategy {
  if (Array.isArray(valA) || Array.isArray(valB) || Array.isArray(ancestor)) {
    return 'or_set';
  }
  if (typeof valA === 'number' && typeof valB === 'number') {
    if (typeof ancestor === 'number') {
      const deltaA = valA - ancestor;
      const deltaB = valB - ancestor;
      if (deltaA >= 0 && deltaB >= 0) {
        return 'g_counter';
      }
    }
    return 'lww_register';
  }
  return 'lww_register';
}

function mergeLwwRegister(
  valA: unknown, valB: unknown, tsA: number, tsB: number
): { value: unknown; source: string } {
  if (tsA >= tsB) {
    return { value: valA, source: 'A' };
  }
  return { value: valB, source: 'B' };
}

function mergeGCounter(valA: number, valB: number, ancestor: number): number {
  // G-Counter merge: ancestor + max(deltaA, deltaB) for each counter
  // In a single-actor model, this reduces to taking the max
  const deltaA = valA - ancestor;
  const deltaB = valB - ancestor;
  return ancestor + Math.max(deltaA, deltaB);
}

function mergeOrSet(
  arrA: unknown[], arrB: unknown[], ancestor: unknown[]
): unknown[] {
  // OR-Set semantics: union of all additions, minus observed removals
  const ancestorSet = new Set(ancestor.map(v => JSON.stringify(v)));
  const addedByA = arrA.filter(v => !ancestorSet.has(JSON.stringify(v)));
  const addedByB = arrB.filter(v => !ancestorSet.has(JSON.stringify(v)));

  const removedByA = ancestor.filter(v => !arrA.map(x => JSON.stringify(x)).includes(JSON.stringify(v)));
  const removedByB = ancestor.filter(v => !arrB.map(x => JSON.stringify(x)).includes(JSON.stringify(v)));

  // Items removed by both sides are truly removed (tombstoned)
  const tombstoned = new Set(
    removedByA
      .filter(v => removedByB.some(rb => JSON.stringify(rb) === JSON.stringify(v)))
      .map(v => JSON.stringify(v))
  );

  // Start with ancestor, remove tombstoned, add new items from both sides
  const base = ancestor.filter(v => !tombstoned.has(JSON.stringify(v)));
  const seen = new Set(base.map(v => JSON.stringify(v)));

  for (const item of [...addedByA, ...addedByB]) {
    const key = JSON.stringify(item);
    if (!seen.has(key)) {
      base.push(item);
      seen.add(key);
    }
  }

  return base;
}

export class CrdtMergeResolverProvider {
  resolve(conflict: Conflict, _config: ResolverConfig): Resolution {
    const ancestor = conflict.ancestor ?? {};
    const tsA = conflict.timestampA ?? 0;
    const tsB = conflict.timestampB ?? 0;
    const allFields = new Set([
      ...Object.keys(conflict.versionA),
      ...Object.keys(conflict.versionB),
      ...Object.keys(ancestor),
    ]);

    const merged: Record<string, unknown> = {};
    const fieldStrategies: Record<string, CrdtStrategy> = {};
    const fieldSources: Record<string, string> = {};

    for (const field of allFields) {
      const valA = conflict.versionA[field];
      const valB = conflict.versionB[field];
      const valAnc = ancestor[field];

      const crdtType = detectFieldCrdtType(valA, valB, valAnc);
      fieldStrategies[field] = crdtType;

      switch (crdtType) {
        case 'lww_register': {
          const result = mergeLwwRegister(valA, valB, tsA, tsB);
          merged[field] = result.value;
          fieldSources[field] = `lww_${result.source.toLowerCase()}`;
          break;
        }
        case 'g_counter': {
          const counterVal = mergeGCounter(
            valA as number,
            valB as number,
            (valAnc as number) ?? 0,
          );
          merged[field] = counterVal;
          fieldSources[field] = 'g_counter_max';
          break;
        }
        case 'or_set': {
          const setVal = mergeOrSet(
            Array.isArray(valA) ? valA : [],
            Array.isArray(valB) ? valB : [],
            Array.isArray(valAnc) ? valAnc : [],
          );
          merged[field] = setVal;
          fieldSources[field] = 'or_set_union';
          break;
        }
      }
    }

    return {
      winner: merged,
      strategy: 'crdt_merge',
      details: {
        entityId: conflict.entityId,
        fieldStrategies,
        fieldSources,
        totalFields: allFields.size,
        convergenceGuaranteed: true,
      },
    };
  }

  canAutoResolve(_conflict: Conflict): boolean {
    return true;
  }
}

export default CrdtMergeResolverProvider;

File: repertoire/data-integration/providers/conflict_resolver/CrdtMerge.swift
// CRDT-based conflict-free merge resolver
// Always auto-resolves mathematically using convergent replicated data types.
// Applies per-field CRDT merge strategies: LWW-Register for scalars,
// G-Counter for numeric increments, OR-Set for collections.

import Foundation

public struct Conflict {
    public let entityId: String
    public let versionA: [String: Any]
    public let versionB: [String: Any]
    public let ancestor: [String: Any]?
    public let fieldConflicts: [String]
    public let timestampA: UInt64?
    public let timestampB: UInt64?
}

public struct ResolverConfig {
    public let options: [String: Any]
}

public struct Resolution {
    public let winner: [String: Any]
    public let strategy: String
    public let details: [String: Any]
}

public enum ResolverError: Error {
    case mergeFailure(String)
}

public let providerID = "crdt_merge"
public let pluginType = "conflict_resolver"

private enum CrdtStrategy: String {
    case lwwRegister = "lww_register"
    case gCounter = "g_counter"
    case orSet = "or_set"
}

private func detectFieldCrdtType(valA: Any?, valB: Any?, ancestor: Any?) -> CrdtStrategy {
    if valA is [Any] || valB is [Any] || ancestor is [Any] {
        return .orSet
    }

    if let numA = valA as? Double, let numB = valB as? Double {
        if let numAnc = ancestor as? Double {
            let deltaA = numA - numAnc
            let deltaB = numB - numAnc
            if deltaA >= 0 && deltaB >= 0 {
                return .gCounter
            }
        }
        return .lwwRegister
    }

    if let numA = valA as? Int, let numB = valB as? Int {
        if let numAnc = ancestor as? Int {
            let deltaA = numA - numAnc
            let deltaB = numB - numAnc
            if deltaA >= 0 && deltaB >= 0 {
                return .gCounter
            }
        }
        return .lwwRegister
    }

    return .lwwRegister
}

private func mergeLwwRegister(valA: Any?, valB: Any?, tsA: UInt64, tsB: UInt64) -> (value: Any?, source: String) {
    if tsA >= tsB {
        return (valA, "lww_a")
    }
    return (valB, "lww_b")
}

private func mergeGCounter(valA: Double, valB: Double, ancestor: Double) -> Double {
    let deltaA = valA - ancestor
    let deltaB = valB - ancestor
    return ancestor + max(deltaA, deltaB)
}

private func anyToString(_ val: Any) -> String {
    return String(describing: val)
}

private func mergeOrSet(arrA: [Any], arrB: [Any], ancestor: [Any]) -> [Any] {
    let ancestorSet = Set(ancestor.map { anyToString($0) })
    let arrAStrs = Set(arrA.map { anyToString($0) })
    let arrBStrs = Set(arrB.map { anyToString($0) })

    // Items removed by both sides are tombstoned
    let removedByA = ancestorSet.subtracting(arrAStrs)
    let removedByB = ancestorSet.subtracting(arrBStrs)
    let tombstoned = removedByA.intersection(removedByB)

    // Start with ancestor minus tombstoned
    var result: [Any] = ancestor.filter { !tombstoned.contains(anyToString($0)) }
    var seen = Set(result.map { anyToString($0) })

    // Add new items from both sides
    for item in arrA + arrB {
        let key = anyToString(item)
        if !seen.contains(key) && !ancestorSet.contains(key) {
            result.append(item)
            seen.insert(key)
        }
    }

    return result
}

private func toDouble(_ val: Any?) -> Double {
    if let d = val as? Double { return d }
    if let i = val as? Int { return Double(i) }
    return 0.0
}

public final class CrdtMergeResolverProvider {
    public init() {}

    public func resolve(conflict: Conflict, config: ResolverConfig) throws -> Resolution {
        let ancestor = conflict.ancestor ?? [:]
        let tsA = conflict.timestampA ?? 0
        let tsB = conflict.timestampB ?? 0

        var allFields = Set<String>()
        for key in conflict.versionA.keys { allFields.insert(key) }
        for key in conflict.versionB.keys { allFields.insert(key) }
        for key in ancestor.keys { allFields.insert(key) }

        var merged: [String: Any] = [:]
        var fieldStrategies: [String: String] = [:]
        var fieldSources: [String: String] = [:]

        for field in allFields.sorted() {
            let valA = conflict.versionA[field]
            let valB = conflict.versionB[field]
            let valAnc = ancestor[field]

            let crdtType = detectFieldCrdtType(valA: valA, valB: valB, ancestor: valAnc)
            fieldStrategies[field] = crdtType.rawValue

            switch crdtType {
            case .lwwRegister:
                let result = mergeLwwRegister(valA: valA, valB: valB, tsA: tsA, tsB: tsB)
                merged[field] = result.value
                fieldSources[field] = result.source

            case .gCounter:
                let numA = toDouble(valA)
                let numB = toDouble(valB)
                let numAnc = toDouble(valAnc)
                merged[field] = mergeGCounter(valA: numA, valB: numB, ancestor: numAnc)
                fieldSources[field] = "g_counter_max"

            case .orSet:
                let a = valA as? [Any] ?? []
                let b = valB as? [Any] ?? []
                let anc = valAnc as? [Any] ?? []
                merged[field] = mergeOrSet(arrA: a, arrB: b, ancestor: anc)
                fieldSources[field] = "or_set_union"
            }
        }

        return Resolution(
            winner: merged,
            strategy: "crdt_merge",
            details: [
                "entityId": conflict.entityId,
                "fieldStrategies": fieldStrategies,
                "fieldSources": fieldSources,
                "totalFields": allFields.count,
                "convergenceGuaranteed": true,
            ]
        )
    }

    public func canAutoResolve(conflict: Conflict) -> Bool {
        return true
    }
}
